File: sliding_window/longest_substring_without_repeating_characters.py
# Q:3 - https://leetcode.com/problems/longest-substring-without-repeating-characters/

class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        map = {}
        longest = 0
        left, right = 0, 0

        while right < len(s):
            if s[right] not in map:
                map[s[right]] = 1
            else:
                # remove the char at left while the char at right is in the map
                while s[right] in map:
                    map.pop(s[left])
                    left += 1
                map[s[right]] = 1

            longest = max(longest, right - left + 1)
            right += 1

        return longest


c = Solution()

s1 = "abcabcbb"
s2 = "bbbbb"
s3 = "pwwkew"
s4 = "dvdf"

print(c.lengthOfLongestSubstring(s1))
print(c.lengthOfLongestSubstring(s2))
print(c.lengthOfLongestSubstring(s3))
print(c.lengthOfLongestSubstring(s4))

File: sliding_window/longest_substring_without_repeating_characters.js
// Q:3 - https://leetcode.com/problems/longest-substring-without-repeating-characters/

/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function (s) {
  let longest = 0,
    start = 0,
    end = 0;
  const map = new Map();

  while (end < s.length) {
    if (!map.has(s[end])) {
      map.set(s[end], true);
    } else {
      while (map.has(s[end])) {
        map.delete(s[start]);
        start++;
      }
      map.set(s[end], true);
    }

    longest = Math.max(longest, end - start + 1);
    end++;
  }

  return longest;
};
